After Upgrade to 15.9.2 Circuclar Dependency Error

### Current Behavior

When linting the project, I get the message, that there is a circular dependency between an app and a library.

`/Users/xxx/apps/jpi/src/pages/p/[page].tsx
   7:1   error    Circular dependency between "jpi" and "ui" detected: jpi -> ui -> jpi

Circular file chain:
- apps/jpi/src/pages/p/[page].tsx
- undefined  @nrwl/nx/enforce-module-boundaries
`

However it is nowhere stated where that circular dependency should come from.

The dep-graph does **not** show any imports from "lib" into "jpi". There is a arrow, but the file list is empty.

### Expected Behavior

I wish it would just work like before the update to 15.9.2

### GitHub Repo

_No response_

### Steps to Reproduce

1. run lint
2. see error

### Nx Report

```shell
Node : 16.15.1
   OS   : darwin arm64
   pnpm : 7.5.0
   
   nx                      : 15.9.2
   @nrwl/js                : 15.9.2
   @nrwl/jest              : 15.9.2
   @nrwl/linter            : 15.9.2
   @nrwl/workspace         : 15.9.2
   @nrwl/cli               : 15.9.2
   @nrwl/cypress           : 15.9.2
   @nrwl/eslint-plugin-nx  : 15.9.2
   @nrwl/next              : 15.9.2
   @nrwl/react             : 15.9.2
   @nrwl/storybook         : 15.9.2
   @nrwl/tao               : 15.9.2
   @nrwl/web               : 15.9.2
   @nrwl/webpack           : 15.9.2
   @nrwl/nx-cloud          : 15.3.5
   typescript              : 4.9.4
```


### Failure Logs

_No response_

### Additional Information

_No response_

Having this one as well, `nx reset` helped but it happened at least twice after it as well...
Might be something bad happening with caching? IIRC, it shouldn't cache failed tasks output, no?


```sh
 2:1  error  Circular dependency between "feature-a" and "feature-b" detected: feature-a -> feature-b -> feature-c -> feature-a

Circular file chain:
- libs/feature-a/Module.ts
- undefined
- [ ... ] # some entries here

@nrwl/nx/enforce-module-boundaries
```

--- a/packages/nx/src/project-graph/build-project-graph.spec.ts
+++ b/packages/nx/src/project-graph/build-project-graph.spec.ts
@@ -232,6 +232,7 @@ describe('project graph', () => {
     expect(graph.dependencies).toEqual({
       api: [{ source: 'api', target: 'npm:express', type: 'static' }],
       demo: [
+        { source: 'demo', target: 'api', type: 'implicit' },
         {
           source: 'demo',
           target: 'ui',
@@ -243,7 +244,6 @@ describe('project graph', () => {
           target: 'lazy-lib',
           type: 'dynamic',
         },
-        { source: 'demo', target: 'api', type: 'implicit' },
       ],
       'demo-e2e': [],
       'lazy-lib': [],

--- a/packages/nx/src/project-graph/project-graph-builder.spec.ts
+++ b/packages/nx/src/project-graph/project-graph-builder.spec.ts
@@ -2,6 +2,7 @@ import { ProjectGraphBuilder } from './project-graph-builder';
 
 describe('ProjectGraphBuilder', () => {
   let builder: ProjectGraphBuilder;
+
   beforeEach(() => {
     builder = new ProjectGraphBuilder();
     builder.addNode({
@@ -208,5 +209,114 @@ describe('ProjectGraphBuilder', () => {
       target: [],
       target2: [],
     });
+  });
+
+  it('should prune dependencies when removing explicit dependencies from the files', () => {
+    builder.addImplicitDependency('source', 'target');
+    builder.addExternalNode({
+      name: 'npm:external',
+      type: 'npm',
+      data: {
+        version: '1.0.0',
+        packageName: 'external',
+      },
+    });
+    builder.addExternalNode({
+      name: 'npm:external2',
+      type: 'npm',
+      data: {
+        version: '1.0.0',
+        packageName: 'external2',
+      },
+    });
+    builder.addStaticDependency('npm:external', 'npm:external2');
+    builder.addStaticDependency('source', 'npm:external', 'source/index.ts');
+    builder.addDynamicDependency('source', 'npm:external2', 'source/second.ts');
+    const graph = builder.getUpdatedProjectGraph();
+
+    expect(graph.dependencies).toMatchInlineSnapshot(`
+      {
+        "npm:external": [
+          {
+            "source": "npm:external",
+            "target": "npm:external2",
+            "type": "static",
+          },
+        ],
+        "source": [
+          {
+            "source": "source",
+            "target": "target",
+            "type": "implicit",
+          },
+          {
+            "source": "source",
+            "target": "npm:external",
+            "type": "static",
+          },
+          {
+            "source": "source",
+            "target": "npm:external2",
+            "type": "dynamic",
+          },
+        ],
+        "target": [],
+      }
+    `);
+    expect(graph.nodes['source'].data.files).toMatchInlineSnapshot(`
+      [
+        {
+          "dependencies": [
+            {
+              "source": "source",
+              "target": "npm:external",
+              "type": "static",
+            },
+          ],
+          "file": "source/index.ts",
+        },
+        {
+          "dependencies": [
+            {
+              "source": "source",
+              "target": "npm:external2",
+              "type": "dynamic",
+            },
+          ],
+          "file": "source/second.ts",
+        },
+      ]
+    `);
+
+    const newBuilder = new ProjectGraphBuilder(graph);
+    // remove static dependency from the file
+    delete newBuilder.graph.nodes['source'].data.files[0].dependencies;
+
+    const updatedGraph = newBuilder.getUpdatedProjectGraph();
+
+    expect(updatedGraph.dependencies).toMatchInlineSnapshot(`
+      {
+        "npm:external": [
+          {
+            "source": "npm:external",
+            "target": "npm:external2",
+            "type": "static",
+          },
+        ],
+        "source": [
+          {
+            "source": "source",
+            "target": "target",
+            "type": "implicit",
+          },
+          {
+            "source": "source",
+            "target": "npm:external2",
+            "type": "dynamic",
+          },
+        ],
+        "target": [],
+      }
+    `);
   });
 });

--- a/packages/nx/src/project-graph/project-graph-builder.ts
+++ b/packages/nx/src/project-graph/project-graph-builder.ts
@@ -385,6 +385,23 @@ export class ProjectGraphBuilder {
         }
       }
     }
+    // cleanup stale dependencies
+    for (const [targetProject, types] of alreadySetTargetProjects.entries()) {
+      const node = this.graph.nodes[sourceProject];
+      if (node) {
+        for (const depType of types.keys()) {
+          if (
+            !node.data.files.some((f) =>
+              f.dependencies?.some(
+                (d) => d.target === targetProject && d.type === depType
+              )
+            )
+          ) {
+            types.delete(depType);
+          }
+        }
+      }
+    }
     return alreadySetTargetProjects;
   }
 }

--- a/packages/nx/src/project-graph/project-graph-builder.ts
+++ b/packages/nx/src/project-graph/project-graph-builder.ts
@@ -284,10 +284,6 @@ export class ProjectGraphBuilder {
     const isDuplicate = !!this.graph.dependencies[sourceProjectName].find(
       (d) => d.target === targetProjectName && d.type === type
     );
-    // do not add duplicate to project
-    if (isDuplicate && !sourceProjectFile) {
-      return;
-    }
 
     const dependency = {
       source: sourceProjectName,
@@ -321,13 +317,11 @@ export class ProjectGraphBuilder {
       ) {
         fileData.dependencies.push(dependency);
       }
+    } else if (!isDuplicate) {
+      // only add to dependencies section if the source file is not specified
+      // and not already added
+      this.graph.dependencies[sourceProjectName].push(dependency);
     }
-
-    if (isDuplicate) {
-      return;
-    }
-
-    this.graph.dependencies[sourceProjectName].push(dependency);
   }
 
   private removeDependenciesWithNode(name: string) {

--- a/packages/nx/src/project-graph/project-graph-builder.ts
+++ b/packages/nx/src/project-graph/project-graph-builder.ts
@@ -88,6 +88,8 @@ export class ProjectGraphBuilder {
     targetProjectName: string,
     sourceProjectFile?: string
   ): void {
+    // internal nodes must provide sourceProjectFile when creating static dependency
+    // externalNodes do not have sourceProjectFile
     if (this.graph.nodes[sourceProjectName] && !sourceProjectFile) {
       throw new Error(`Source project file is required`);
     }
@@ -110,6 +112,7 @@ export class ProjectGraphBuilder {
     if (this.graph.externalNodes[sourceProjectName]) {
       throw new Error(`External projects can't have "dynamic" dependencies`);
     }
+    // dynamic dependency is always bound to a file
     if (!sourceProjectFile) {
       throw new Error(`Source project file is required`);
     }
@@ -380,6 +383,9 @@ export class ProjectGraphBuilder {
     if (this.graph.dependencies[sourceProject]) {
       const removed = this.removedEdges[sourceProject];
       for (const d of this.graph.dependencies[sourceProject]) {
+        // static and dynamic dependencies of internal projects
+        // will be rebuilt based on the file dependencies
+        // we only need to keep the implicit dependencies
         if (d.type === DependencyType.implicit && !removed?.has(d.target)) {
           if (!alreadySetTargetProjects.has(d.target)) {
             alreadySetTargetProjects.set(d.target, new Map([[d.type, d]]));

--- a/packages/nx/src/project-graph/project-graph-builder.ts
+++ b/packages/nx/src/project-graph/project-graph-builder.ts
@@ -380,29 +380,11 @@ export class ProjectGraphBuilder {
     if (this.graph.dependencies[sourceProject]) {
       const removed = this.removedEdges[sourceProject];
       for (const d of this.graph.dependencies[sourceProject]) {
-        if (!removed || !removed.has(d.target)) {
+        if (d.type === DependencyType.implicit && !removed?.has(d.target)) {
           if (!alreadySetTargetProjects.has(d.target)) {
             alreadySetTargetProjects.set(d.target, new Map([[d.type, d]]));
           } else {
             alreadySetTargetProjects.get(d.target).set(d.type, d);
-          }
-        }
-      }
-    }
-    // cleanup stale dependencies
-    for (const [targetProject, types] of alreadySetTargetProjects.entries()) {
-      const node = this.graph.nodes[sourceProject];
-      if (node) {
-        for (const depType of types.keys()) {
-          if (
-            depType !== 'implicit' &&
-            !node.data.files.some((f) =>
-              f.dependencies?.some(
-                (d) => d.target === targetProject && d.type === depType
-              )
-            )
-          ) {
-            types.delete(depType);
           }
         }
       }

--- a/packages/nx/src/project-graph/project-graph-builder.ts
+++ b/packages/nx/src/project-graph/project-graph-builder.ts
@@ -311,7 +311,11 @@ export class ProjectGraphBuilder {
       if (!fileData.dependencies) {
         fileData.dependencies = [];
       }
-      if (!fileData.dependencies.find((t) => t.target === targetProjectName)) {
+      if (
+        !fileData.dependencies.find(
+          (t) => t.target === targetProjectName && t.type === type
+        )
+      ) {
         fileData.dependencies.push(dependency);
       }
     }
@@ -391,6 +395,7 @@ export class ProjectGraphBuilder {
       if (node) {
         for (const depType of types.keys()) {
           if (
+            depType !== 'implicit' &&
             !node.data.files.some((f) =>
               f.dependencies?.some(
                 (d) => d.target === targetProject && d.type === depType

